### PR TITLE
interlocked_hl rising from the ashes

### PR DESCRIPTION
**Pull request by dcristoloveanu**
interlocked_hl rising from the ashes

### CODE REVIEW

File: src/sm.c [hunk 1]
```diff
@@ -120,7 +121,7 @@ int sm_open_begin(SM_HANDLE sm)
     else
     {
         /*Codes_SRS_SM_02_039: [ If the state is not SM_CREATED then sm_open_begin shall return SM_EXEC_REFUSED. ]*/
-        LONG state = InterlockedAdd(&sm->state, 0);
+        LONG state = interlocked_add(&sm->state, 0);
```
Comment (mattdurak) [inline src/sm.c:124]: >LONG [](start = 8, length = 4)

int32_t

File: src/sm.c [hunk 2]
```diff
@@ -154,7 +155,7 @@ void sm_open_end(SM_HANDLE sm, bool success)
     else
     {
         /*Codes_SRS_SM_02_041: [ If state is not SM_OPENING then sm_open_end shall return. ]*/
-        LONG state = InterlockedAdd(&sm->state, 0);
+        LONG state = interlocked_add(&sm->state, 0);
```
Comment (mattdurak) [inline src/sm.c:158]: >LONG [](start = 8, length = 4)

int32_t

Comment (mattdurak) [reply src/sm.c:158]: here and everywhere probably

---
In reply to: [501161065](https://github.com/Azure/azure-c-util/pull/66#discussion_r501161065) [](ancestors = 501161065)

**Comment by mattdurak** [comment]
>                 Sleep(1);

Looks like the only reason this file includes windows.h still (besides the LONG type). I created task #8516433 FYI

---
Refers to: src/sm.c:243 in 11d1a7f. [](commit_id = 11d1a7fb2ebaff6b638f6c1c81af7614639ceff4, deletion_comment = False)

File: src/sm.c [hunk 3]
```diff
@@ -250,7 +251,7 @@ static SM_RESULT sm_close_begin_internal(SM_HANDLE sm)
         } while (1);
 
         /*Codes_SRS_SM_02_053: [ sm_close_begin shall set SM_CLOSE_BIT to 0. ]*/
-        (void)InterlockedAnd(&sm->state, ~(ULONG)SM_CLOSE_BIT);
+        (void)interlocked_and(&sm->state, ~(ULONG)SM_CLOSE_BIT);
```
Comment (mattdurak) [inline src/sm.c:254]: >ULONG [](start = 44, length = 5)

uint32_t

File: tests/sm_int/sm_int.c [hunk 4]
```diff
@@ -17,7 +17,7 @@
 #include "azure_macro_utils/macro_utils.h"
 
 #include "azure_c_pal/timer.h"
-#include "azure_c_pal/interlocked_hl.h"
+#include "azure_c_util/interlocked_hl.h"
```
Comment (mattdurak) [inline tests/sm_int/sm_int.c:20]: >interlocked_hl [](start = 23, length = 14)

not used here anyway?

**Comment by mattdurak** [comment]
> #include "umock_c/umocktypes_windows.h"

Shouldn't be needed

---
Refers to: tests/sm_ut/sm_ut.c:17 in 11d1a7f. [](commit_id = 11d1a7fb2ebaff6b638f6c1c81af7614639ceff4, deletion_comment = False)

**Comment by mattdurak** [comment]
>                     if (InterlockedHL_WaitForValue(&sm->non_barrier_call_count, 0, INFINITE) != INTERLOCKED_HL_OK)

UINT32_MAX

---
Refers to: src/sm.c:221 in 11d1a7f. [](commit_id = 11d1a7fb2ebaff6b638f6c1c81af7614639ceff4, deletion_comment = False)

File: src/sm.c [hunk 5]
```diff
@@ -221,13 +222,13 @@ static SM_RESULT sm_close_begin_internal(SM_HANDLE sm)
                     {
                         /*Codes_SRS_SM_02_071: [ If there are any failures then sm_close_begin shall fail and return SM_ERROR. ]*/
                         LogError("sm name=%s. failure in InterlockedHL_WaitForValue(&sm->non_barrier_call_count=%p, 0, INFINITE), state was %" PRI_SM_STATE "", sm->name, &sm->non_barrier_call_count, SM_STATE_VALUE(state));
```
Comment (mattdurak) [inline src/sm.c:224]: >INFINITE [](start = 119, length = 8)

UINT32_MAX

File: src/sm.c [hunk 6]
```diff
@@ -333,14 +334,14 @@ SM_RESULT sm_exec_begin(SM_HANDLE sm)
...
 
             /*Codes_SRS_SM_02_057: [ If the state changed after incrementing n then sm_exec_begin shall return SM_EXEC_REFUSED. ]*/
             if (state1 != state2)
             {
                 LogError("sm name=%s. state changed meanwhile from %" PRI_SM_STATE " to %" PRI_SM_STATE "", sm->name, SM_STATE_VALUE(state1), SM_STATE_VALUE(state2));
-                LONG n = InterlockedDecrement(&sm->non_barrier_call_count);
+                LONG n = interlocked_decrement(&sm->non_barrier_call_count);
                 if (n == 0)
                 {
                     WakeByAddressSingle((void*)&sm->non_barrier_call_count);
```
Comment (mattdurak) [inline src/sm.c:347]: >WakeByAddressSingle [](start = 20, length = 19)

wake_by_address_single (while we're changing the rest of this)?

**Review by mattdurak — APPROVED** [review]
:shipit:

File: src/sm.c (hunk 1 above)
Comment (dcristoloveanu) [reply src/sm.c:124]: Fixed.

---
In reply to: [501160920](https://github.com/Azure/azure-c-util/pull/66#discussion_r501160920) [](ancestors = 501160920)

File: src/sm.c (hunk 2 above)
Comment (dcristoloveanu) [reply src/sm.c:158]: Fixed.

---
In reply to: [501161280](https://github.com/Azure/azure-c-util/pull/66#discussion_r501161280) [](ancestors = 501161280,501161065)

File: src/sm.c (hunk 3 above)
Comment (dcristoloveanu) [reply src/sm.c:254]: Fixed.

---
In reply to: [501163534](https://github.com/Azure/azure-c-util/pull/66#discussion_r501163534) [](ancestors = 501163534)

File: tests/sm_int/sm_int.c (hunk 4 above)
Comment (dcristoloveanu) [reply tests/sm_int/sm_int.c:20]: Removed

---
In reply to: [501164641](https://github.com/Azure/azure-c-util/pull/66#discussion_r501164641) [](ancestors = 501164641)

File: src/sm.c (hunk 5 above)
Comment (dcristoloveanu) [reply src/sm.c:224]: Fixed.

---
In reply to: [501166222](https://github.com/Azure/azure-c-util/pull/66#discussion_r501166222) [](ancestors = 501166222)

**Comment by dcristoloveanu** [comment]
>                     if (InterlockedHL_WaitForValue(&sm->non_barrier_call_count, 0, INFINITE) != INTERLOCKED_HL_OK)

Fixed.

---
In reply to: [705065698](https://github.com/Azure/azure-c-util/pull/66#issuecomment-705065698) [](ancestors = 705065698)

---
Refers to: src/sm.c:221 in 11d1a7f. [](commit_id = 11d1a7fb2ebaff6b638f6c1c81af7614639ceff4, deletion_comment = False)

File: src/sm.c (hunk 6 above)
Comment (dcristoloveanu) [reply src/sm.c:347]: Will do a separate PR, I want to not leave InterlockedHL hanging in the nether

---
In reply to: [501166830](https://github.com/Azure/azure-c-util/pull/66#discussion_r501166830) [](ancestors = 501166830)

**Comment by dcristoloveanu** [comment]
> #include "umock_c/umocktypes_windows.h"

Unneeded thingie removed

---
In reply to: [705064646](https://github.com/Azure/azure-c-util/pull/66#issuecomment-705064646) [](ancestors = 705064646)

---
Refers to: tests/sm_ut/sm_ut.c:17 in 11d1a7f. [](commit_id = 11d1a7fb2ebaff6b638f6c1c81af7614639ceff4, deletion_comment = False)

**Pull request merged**